extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -0,0 +1,1266 @@
+          taskGroup.tasks.put(taskId, new TaskData());
+        }
+        return;
+      }
+    }
+
+    log.info("Creating new pending completion task group for discovered task [%s]", taskId);
+    TaskGroup newTaskGroup = new TaskGroup(startingPartitions);
+    newTaskGroup.tasks.put(taskId, new TaskData());
+    newTaskGroup.completionTimeout = DateTime.now().plus(ioConfig.getCompletionTimeout());
+
+    taskGroupList.add(newTaskGroup);
+  }
+
+  private void updateTaskStatus()
+  {
+    // update status (and startTime if unknown) of current tasks in taskGroups
+    for (TaskGroup group : taskGroups.values()) {
+      for (Map.Entry<String, TaskData> entry : group.tasks.entrySet()) {
+        String taskId = entry.getKey();
+        TaskData taskData = entry.getValue();
+
+        if (taskData.startTime == null) {
+          try {
+            taskData.startTime = getTaskStartTime(taskId);
+            if (taskData.startTime != null) {
+              long millisRemaining = ioConfig.getTaskDuration().getMillis() - (System.currentTimeMillis()
+                                                                               - taskData.startTime.getMillis());
+              if (millisRemaining > 0) {
+                scheduledExec.schedule(buildRunTask(), millisRemaining, TimeUnit.MILLISECONDS);
+              }
+            }
+          }
+          catch (Exception e) {
+            log.warn(e, "Task [%s] failed to return start time, killing task", taskId);
+            killTask(taskId);
+          }
+        }
+
+        taskData.status = taskStorage.getStatus(taskId).get();
+      }
+    }
+
+    // update status of pending completion tasks in pendingCompletionTaskGroups
+    for (List<TaskGroup> taskGroups : pendingCompletionTaskGroups.values()) {
+      for (TaskGroup group : taskGroups) {
+        for (Map.Entry<String, TaskData> entry : group.tasks.entrySet()) {
+          entry.getValue().status = taskStorage.getStatus(entry.getKey()).get();
+        }
+      }
+    }
+  }
+
+  private void checkTaskDuration()
+  {
+    Iterator<Map.Entry<Integer, TaskGroup>> i = taskGroups.entrySet().iterator();
+    while (i.hasNext()) {
+      Map.Entry<Integer, TaskGroup> groupEntry = i.next();
+      Integer groupId = groupEntry.getKey();
+      TaskGroup group = groupEntry.getValue();
+
+      // find the longest running task from this group
+      DateTime earliestTaskStart = DateTime.now();
+      for (TaskData taskData : group.tasks.values()) {
+        if (earliestTaskStart.isAfter(taskData.startTime)) {
+          earliestTaskStart = taskData.startTime;
+        }
+      }
+
+      // if this task has run longer than the configured duration, signal all tasks in the group to persist
+      if (earliestTaskStart.plus(ioConfig.getTaskDuration()).isBeforeNow()) {
+        log.info("Task group [%d] has run for [%s]", groupId, ioConfig.getTaskDuration());
+        Map<Integer, Long> endOffsets = signalTasksToFinish(groupId);
```

Comment:
The trigger for a task being "done" is based on the startTime reported by the task itself (it's "cached" in the supervisor's state but is strictly based on data from the task), so if the supervisor is in the middle of processing tasks that should be finished and then crashes, when the supervisor restarts and newly discovers the currently running tasks, if the tasks match the supervisor's spec and starting offset it'll start tracking the task, then it'll grab the start time from the task and then immediately realize that the task has run for the expected duration and will signal the tasks to finish or whatever it has to do.

If the supervisor crashed after issuing a pause to the task but before issuing a set-end-offsets-and-resume, the next supervisor would try to pause it again (which would succeed) and then continue to calculate the end offsets and start the publishing.

If it did get as far as starting the publishing and then crashed, the task state would have switched from READING to PUBLISHING and then when the next supervisor starts, when discoverTasks() runs and picks up those tasks, it'll check if it's in PUBLISHING state and if so it'll add it to the pendingCompletionTaskGroups of the new supervisor and track it as a publishing task.